File: scripts/SwitchPublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import SwitchCommand

class SwitchPublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", SwitchCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", SwitchCommand, queue_size=10)

        while not rospy.is_shutdown():

            self.message = SwitchCommand()

            self.message.command = "ON"
            
            self.message.isnull = False

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: command = %s" % (self.message.item, rospy.get_time(), self.message.command)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("SwitchPublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    switchPublisher = SwitchPublisher("testSwitch")

    try:
        switchPublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/SwitchPublisher.cpp
#include <ros/ros.h>
#include <openhab_msgs/SwitchCommand.h>

class SwitchPublisher
{
public:
    SwitchPublisher(std::string item_name)
    {
        item_name_ = item_name;
        pub_ = nh_.advertise<openhab_msgs::SwitchCommand>("/openhab/items/" + item_name_ + "/command", 10);
        rate_ = ros::Rate(10); // 10hz
        enable_ = false;
    }

    void start()
    {
        // Turn on publisher.
        enable_ = true;
        pub_ = nh_.advertise<openhab_msgs::SwitchCommand>("/openhab/items/" + item_name_ + "/command", 10);

        while (ros::ok())
        {
            openhab_msgs::SwitchCommand message;
            message.command = "ON";
            message.isnull = false;
            message.header.stamp = ros::Time::now();
            message.header.frame_id = "/base_link";
            message.item = item_name_;

            std::string log_message = "Publishing to " + message.item + " at " + std::to_string(ros::Time::now().toSec()) + ": command = " + message.command;
            ROS_INFO_STREAM(log_message);

            pub_.publish(message);
            rate_.sleep();
        }
    }

    void stop()
    {
        // Turn off publisher.
        enable_ = false;
        pub_.shutdown();
    }

private:
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Rate rate_;
    std::string item_name_;
    bool enable_;
};

int main(int argc, char **argv)
{
    // Initialize the node and name it.
    ros::init(argc, argv, "SwitchPublisherNode");
    // Go to class functions that do all the heavy lifting.
    SwitchPublisher switch_publisher("testSwitch");

    try
    {
        switch_publisher.start();
    }
    catch (ros::Exception &e)
    {
        ROS_ERROR_STREAM("ROS Exception: " << e.what());
    }

    // Allow ROS to go to all callbacks.
    // spin() simply keeps c++ from exiting until this node is stopped
    ros::spin();

    return 0;
}
